src/libraries/System.Net.Primitives/src/System/Net/IPAddress.cs

```diff
@@ -695,18 +684,19 @@ internal bool Equals(IPAddress comparand)
             if (IsIPv6)
             {
                 // For IPv6 addresses, we must compare the full 128-bit representation and the scope IDs.
-                ReadOnlySpan<byte> thisNumbers = MemoryMarshal.AsBytes<ushort>(_numbers);
-                ReadOnlySpan<byte> comparandNumbers = MemoryMarshal.AsBytes<ushort>(comparand._numbers);
-                return
-                    MemoryMarshal.Read<ulong>(thisNumbers) == MemoryMarshal.Read<ulong>(comparandNumbers) &&
-                    MemoryMarshal.Read<ulong>(thisNumbers.Slice(sizeof(ulong))) == MemoryMarshal.Read<ulong>(comparandNumbers.Slice(sizeof(ulong))) &&
-                    PrivateScopeId == comparand.PrivateScopeId;
-            }
-            else
-            {
-                // For IPv4 addresses, compare the integer representation.
-                return comparand.PrivateAddress == PrivateAddress;
+                // We give JIT a hint that the arrays are always of length IPv6AddressShorts, so it
```

Comment:
Does `Debug.Assert` have a side effect on JIT decisions? Even in `Release` builds? Is it mentioned somewhere or is it not guaranteed? What if the assertion no longer holds in `Release`?

Reply:
> Does `Debug.Assert` have a side effect on JIT decisions? Even in `Release` builds? Is it mentioned somewhere or is it not guaranteed? What if the assertion no longer holds in `Release`?

Debug.Assert doesn't exist in Release code, it's Debug only

Reply:
Thank you for clarification. I misunderstood your code comment, and thus the question arose.